### DISCUSSION TITLE
Does extracting into function work?

### DISCUSSION THREAD

**Discussion by mb720** [mrcjkb/rustaceanvim#513] [Q&A]
Hi and thanks for rustaceanvim!

I'm wondering if extracting code blocks into a function currently works in general. Currently, I get "No code actions available" when selecting a few lines and running `:RustLsp codeAction` or `:lua vim.lsp.buf.code_action()`.

For example, trying to extract the visually selected for loop in this program with `:RustLsp codeAction` results in "No code actions available":

    fn main() {
        println!("Hello, world!");
        for i in 0..10 {
            println!("I'm counting {}", i);
        }
    }

Other code actions do work: Placing the cursor on the `in` of the loop and calling `:RustLsp codeAction` produces "Replace this for loop with `Iterator::for_each`".

---

`:RustLsp logFile` shows no output after visually selecting lines and calling `:RustLsp codeAction`.

Output of `:checkhealth rustaceanvim`:

    rustaceanvim: require("rustaceanvim.health").check()

    Checking for Lua dependencies ~
    - WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

    Checking external dependencies ~
    - OK rust-analyzer: found rust-analyzer 1.81.0 (eeb90cd 2024-09-04)
    - OK Cargo: found cargo 1.81.0 (2dbb1af80 2024-08-20)
    - OK rustc: found rustc 1.81.0 (eeb90cda1 2024-09-04)

    Checking config ~
    - OK No errors found in config.

    Checking for conflicting plugins ~
    - OK No conflicting plugins detected.

    Checking for tree-sitter parser ~
    - WARNING No tree-sitter parser for Rust detected. Required by 'Rustc unpretty' command.

**Comment by mrcjkb** [comment]
It should be possible. If it's not working, it's likely a rust-analyzer issue.

See also:

- https://github.com/rust-lang/rust-analyzer/issues/17280
- https://github.com/rust-lang/rust-analyzer/issues/11577
- [This comment](https://github.com/rust-lang/rust-analyzer/issues/11577#issuecomment-2307179961), regarding rustaceanvim specifically.
